src/Microsoft.ComponentDetection.Detectors/nuget/NuGetMSBuildBinaryLogComponentDetector.cs

```diff
@@ -0,0 +1,212 @@
+    {
+        var doRemoveOperation = node is RemoveItem;
+        var doAddOperation = node is AddItem;
+        if (TopLevelPackageItemNames.Contains(node.Name))
+        {
+            var projectEvaluation = node.GetNearestParent<ProjectEvaluation>();
+            if (projectEvaluation is not null)
+            {
+                foreach (var child in node.Children.OfType<Item>())
+                {
+                    var packageName = child.Name;
+                    if (!topLevelDependencies.TryGetValue(projectEvaluation.ProjectFile, out var topLevel))
+                    {
+                        topLevel = new(StringComparer.OrdinalIgnoreCase);
+                        topLevelDependencies[projectEvaluation.ProjectFile] = topLevel;
+                    }
+
+                    if (doRemoveOperation)
+                    {
+                        topLevel.Remove(packageName);
+                    }
+
+                    if (doAddOperation)
+                    {
+                        topLevel.Add(packageName);
+                    }
+                }
+            }
+        }
+        else if (ResolvedPackageItemNames.TryGetValue(node.Name, out var metadataNames))
+        {
+            var nameMetadata = metadataNames.NameMetadata;
+            var versionMetadata = metadataNames.VersionMetadata;
+            var originalProject = node.GetNearestParent<Project>();
+            if (originalProject is not null)
+            {
+                foreach (var child in node.Children.OfType<Item>())
+                {
+                    var packageName = GetChildMetadataValue(child, nameMetadata);
+                    var packageVersion = GetChildMetadataValue(child, versionMetadata);
+                    if (packageName is not null && packageVersion is not null)
+                    {
+                        var project = originalProject;
+                        while (project is not null)
+                        {
+                            if (!projectResolvedDependencies.TryGetValue(project.ProjectFile, out var projectDependencies))
+                            {
+                                projectDependencies = new(StringComparer.OrdinalIgnoreCase);
+                                projectResolvedDependencies[project.ProjectFile] = projectDependencies;
+                            }
+
+                            if (doRemoveOperation)
+                            {
+                                projectDependencies.Remove(packageName);
+                            }
+
+                            if (doAddOperation)
+                            {
+                                projectDependencies[packageName] = packageVersion;
+                            }
+
+                            project = project.GetNearestParent<Project>();
```

Comment:
hmm, why are you walking up the project chain? I don't think these items flow to the calling project?

Reply:
I found I needed this to track transitive dependencies.  E.g., if `library.csproj` has `<PackageReference Include="Some.Package" />` and `unitTests.csproj` has `<ProjectReference Include="..\library\library.csproj" />` this will add `Some.Package` first to `library.csproj` and then crawl up the chain to `unitTests.csproj` so that the dependency is properly reported for both projects.  The only oddity with this (and I may have to special case it) is the `.sln` file also appears in a `Project` node, but I _do_ seem to be getting the proper hierarchy.

Reply:
I don't think I understand this. `unitTests.csproj` should have `Some.Package` in its assets file and pull assets out of it in its `ResolvePackageAssets`. Can you share a log or project setup where this is necessary?

Reply:
I'm saying that the targets that read the assets file should produce `Some.Package` in the transitive case, so walking the project graph here doesn't make sense to me.

Reply:
> ...that read the assets file...

That's the problem, the assets file could be wrong, so I need to crawl it manually.  The PR description explains a scenario where the assets file isn't correct.

Reply:
The assets file will still be a superset of what the project might use.  You get the "real usage" by looking at the outcome of package resolution / conflict resolution.   

If you walk the project graph you end up trying to replay parts of the build.  We shouldn't do that when we can just observe what it did.

Also - the assets file isn't "Incorrect" here - it is correct from NuGet's perspective.  It's just that the build has more policy that it applies to decide if it will actually use a package's contents.  If that package contributes assets which are "older" than some other contribution it will be dropped.  The most common case is framework, as you mentioned -- it's what we designed the feature for in the SDK.  Technically it could be any conflict though - when any two packages try to provide the same file the build will compare them to decide who's copy wins.